src/java/htsjdk/samtools/sra/SRAAlignmentIterator.java

```diff
@@ -0,0 +1,198 @@
+*  purpose.
+*
+*  Please cite the author in any work or product based on this material.
+*
+* ===========================================================================
+*
+*/
+
+package htsjdk.samtools.sra;
+
+import htsjdk.samtools.*;
+import ngs.*;
+
+import java.util.ArrayList;
+import java.util.Iterator;
+import java.util.List;
+
+/**
+ * Created by andrii.nikitiuk on 9/3/15.
+ */
+public class SRAAlignmentIterator implements Iterator<SAMRecord> {
+    private ValidationStringency validationStringency;
+
+    private ReadCollection run;
+    private SAMFileHeader header;
+    private final List<Reference> cachedReferences;
+    private List<Long> referencesLengths;
+    private Iterator<Chunk> referencesChunksIterator;
+    private int currentReference = -1;
+
+    private boolean hasMoreReferences = true;
+
+    private AlignmentIterator alignedIterator;
+    private Boolean hasMoreAlignments = false;
+
+    private SRALazyRecord lastRecord;
+
+    public static List<Long> getReferencesLengths(ReadCollection run) throws ErrorMsg {
+        ReferenceIterator refIt = run.getReferences();
+        List<Long> lengths = new ArrayList<Long>();
+        while (refIt.nextReference()) {
+            long refLen = refIt.getLength();
+            // lets optimize references so they always align in 5000 bases positions
+            if (refLen % 5000 != 0) {
+                refLen += 5000 - (refLen % 5000);
+            }
+            lengths.add(refLen);
+        }
+        return lengths;
+    }
+
+    public SRAAlignmentIterator(final ReadCollection run, final SAMFileHeader header, final List<Reference> cachedReferences,
+                                final SRAIterator.RecordRangleInfo recordRangleInfo, final Chunk chunk) {
+        this.run = run;
+        this.header = header;
+        this.cachedReferences = cachedReferences;
+        this.referencesLengths = recordRangleInfo.getReferenceLengths();
+
+        List<Chunk> referencesChunks = new ArrayList<Chunk>();
+        long refOffset = 0;
+        for (Long refLen : referencesLengths) {
+            // TODO: make sure borders are checked correctly
+            if (chunk.getChunkStart() - refOffset >= refLen || chunk.getChunkEnd() - refOffset <= 0) {
+                referencesChunks.add(null);
+            } else {
+                long refChunkStart = chunk.getChunkStart() - refOffset;
+                if (refChunkStart < 0) {
+                    refChunkStart = 0;
+                }
+
+                long refChunkEnd = chunk.getChunkEnd() - refOffset;
+                if (refChunkEnd > refLen) {
+                    refChunkEnd = refLen;
+                }
+                referencesChunks.add(new Chunk(refChunkStart, refChunkEnd));
+            }
+
+            refOffset += refLen;
+        }
+        referencesChunksIterator = referencesChunks.iterator();
+
+        try {
+            nextReference();
+        } catch (final Exception e) {
+            throw new RuntimeException(e);
```

Comment:
Can you add an explanatory error message here explaining to the client what went wrong?